reference/FungibleToken.scilla

```diff
@@ -0,0 +1,429 @@
+scilla_version 0
+
+(***************************************************)
+(*               Associated library                *)
+(***************************************************)
+import BoolUtils ListUtils IntUtils
+library FungibleToken
+
+let one_msg = 
+  fun (msg : Message) => 
+  let nil_msg = Nil {Message} in
+  Cons {Message} msg nil_msg
+
+(* Error events *)
+type Error =
+  | CodeNotAuthorized
+  | CodeNotFound
+  | CodeNotAllowed
+
+let makeError =
+  fun (result : Error) =>
+    let result_code = 
+      match result with
+      | CodeNotAuthorized    => Int32 -1
+      | CodeNotFound         => Int32 -2
+      | CodeNotAllowed      => Int32 -3
+      end
+    in
+    { _eventname : "Error"; code : result_code }
+
+
+let le_int =
+  fun (a : Uint128) => fun (b : Uint128) =>
+    let x = builtin lt a b in
+    match x with
+    | True => True
+    | False =>
+      let y = builtin eq a b in
+      match y with
+      | True => True
+      | False => False
+      end
+    end
+
+let min_int =
+  fun (a : Uint128) => fun (b : Uint128) =>
+  let min = builtin lt a b in
+  match min with
+  | True =>
+    a
+  | False =>
+    b
+  end
+    
+(* A util function to test equality *)
+let f_eq =
+  fun (a : ByStr20) =>
+  fun (b : ByStr20) =>
+    builtin eq a b
+
+(* Instantiate a type function to test membership in a list *)
+let is_default_operator = @list_mem ByStr20
+    
+
+(***************************************************)
+(*             The contract definition             *)
+(***************************************************)
+
+contract FungibleToken
+(contract_owner: ByStr20,
+  name : String,
+  symbol: String,
+  decimals: Uint32,
+  default_operators : List ByStr20
+)
+
+(* Mutable fields *)
+
+field total_tokens : Uint128 = Uint128 0
+
+field balances_map: Map ByStr20 Uint128 
+  = Emp ByStr20 Uint128
+
+field operators_map: Map ByStr20 (Map ByStr20 Bool) 
+  = Emp ByStr20 (Map ByStr20 Bool)
+
+field revoked_default_operators : Map ByStr20 (Map ByStr20 Bool) 
+  = Emp ByStr20 (Map ByStr20 Bool)
+
+field allowances_map: Map ByStr20 (Map ByStr20 Uint128) 
+  = Emp ByStr20 (Map ByStr20 Uint128)
+
+(* Procedures *)
+
+(* Emit Errors *)
+procedure MakeError(err : Error)
```

Comment:
Note that this doesn't actually emit a Scilla error (`throw`), so maybe we can call it `MakeErrorEvent` ?